Refresh.GameServer/Services/OAuth/OAuthClient.cs

```diff
@@ -0,0 +1,232 @@
+    protected abstract string TokenRevocationEndpoint { get; }
+    public abstract bool TokenRevocationSupported { get; }
+    
+    protected abstract string ClientId { get; }
+    protected abstract string ClientSecret { get; }
+    
+    protected abstract string RedirectUri { get; }
+
+    protected OAuthClient(Logger logger)
+    {
+        this.Logger = logger;
+    }
+
+    public virtual void Initialize()
+    {
+        this.Client = new HttpClient();
+        
+        this.Client.DefaultRequestHeaders.Accept.Clear();
+        // Explicitly mark that we want JSON responses, as some servers (GitHub) will return URL encoded text instead by default
+        this.Client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        this.Client.DefaultRequestHeaders.UserAgent.Clear();
+        // Default user agent header
+        this.Client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Refresher", VersionInformation.Version));
+    }
+
+    /// <summary>
+    /// Constructs a URL to send a user to, which they use to authorize Refresh
+    /// </summary>
+    /// <param name="state">The `state` parameter of the authorization</param>
+    /// <returns>The authorization URL to redirect the user to</returns>
+    public abstract string GetOAuthAuthorizationUrl(string state);
+    
+    /// <summary>
+    /// Acquires an access and refresh token using the provided authorization code
+    /// </summary>
+    /// <param name="authCode">The authorization code</param>
+    /// <returns>The acquired access and refresh tokens</returns>
+    public OAuth2AccessTokenResponse AcquireTokenFromAuthorizationCode(string authCode)
+    {
+        HttpResponseMessage result = this.Client.PostAsync(this.TokenEndpoint, new FormUrlEncodedContent([
+            new KeyValuePair<string, string>("grant_type", "authorization_code"),
+            new KeyValuePair<string, string>("code", authCode),
+            new KeyValuePair<string, string>("redirect_uri", this.RedirectUri),
+            new KeyValuePair<string, string>("client_id", this.ClientId),
+            new KeyValuePair<string, string>("client_secret", this.ClientSecret),
+        ])).Result;
+
+        // https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+        if (result.StatusCode == BadRequest)
+        {
+            OAuth2ErrorResponse errorResponse = result.Content.ReadAsJson<OAuth2ErrorResponse>()!;
+
+            throw new Exception($"Unexpected error {errorResponse.Error} when acquiring token! Description: {errorResponse.ErrorDescription}, URI: {errorResponse.ErrorUri}");
+        }
+
+        if (!result.IsSuccessStatusCode)
+            throw new Exception($"Acquiring token failed, server returned status code {result.StatusCode}");
+
+        if ((result.Content.Headers.ContentLength ?? 0) == 0)
+            throw new Exception("Acquiring token failed, request returned no response");
```

Comment:
I get the idea, but the server could stream the request and cause this to fail because there would be no `ContentLength`. Best to just omit this check and let the `ReadAsJson` call fail.